Addition of ExponentialFamily

vishwakftw opened the pull request:
This adds `ExponentialFamily`. `ExponentialFamily` uses concepts of Bregman divergence of the log normalizers of distributions to compute entropy and KL-divergences.

For distributions belonging to an exponential family, this PR adds tests to check implementations of entropy and KL-Divergence functions with both the Monte Carlo and Bregman divergences.

Joint work with @fritzo . Math checked and approved by @fritzo .

vishwakftw committed:
Start ExponentialFamily

vishwakftw committed:
Add initial changes to Normal

vishwakftw committed:
Add lazy_property decorator and exp-dist natural params and log normalizer

vishwakftw committed:
Add testing framework for ExponentialFamily entropy

fritzo committed:
Fix batching bug in ExponentialFamily.entropy()

vishwakftw committed:
Add KL Divergence calculation using Bregman Div and tests

vishwakftw committed:
Add Bernoulli and remove try..except in favour of if..else

vishwakftw committed:
Add support for natural params which are fixed and Pareto distribution

vishwakftw committed:
Add Beta, Binomial, Dirichlet in ExponentialFamily, make natural_params a lazy property with no implicit calls

vishwakftw committed:
Merge branch 'master' into exp-family

vishwakftw committed:
Add bug fixes for tests, still some tests are causing issues

vishwakftw committed:
Fix Normal test

fritzo committed:
Fix kl divergence tests

vishwakftw committed:
Add docstring for ExponentialFamily

vishwakftw committed:
Remove file formatting caused by ^M, add note for ExponentialFamily and fix nit in test_distributions

vishwakftw committed:
Remove redundant mean_carrier_measure information

vishwakftw committed:
Remove redundant .entropy in Binomial, fix flake8 errors

fritzo commented:
I've checked math in: `ExponentialFamily.entropy()`, `ExponentialFamily.kl_divergence()`, and a couple of the `.natural_params` and `.log_normalizer()` implementations.

This PR is mainly for testing: it provides alternative implementations for `.entropy()` and `kl_divergence()` and tests that our two implementations agree; it thus provides higher assurance of correctness than our existing Monte Carlo tests. The implementations in this PR are not used by default in any of our existing distributions.

apaszke commented on torch/distributions/dirichlet.py:
```diff
@@ -80,3 +80,12 @@ def entropy(self):
+    @lazy_property
+    def natural_params(self):
+        V1 = Variable(self.concentration.data, requires_grad=True)
+        return (V1, )
```
Do we plan on having those methods be part of the public interface? They're doing things that aren't fully legal in autograd (rewrapping data in other Variables like this). The semantics are also unclear for me (why does the result always require grad)?

fritzo replied:
I suppose these could be private, since I don't know of an application outside of `.entropy()` and `.kl_divergence()`; however there may be future applications.

The `.natural_params` need to set `requires_grad=True` internally because `.entropy()` and `.kl_divergence` both use `grad()` under the hood, even if the original variables do not require grad. Would you recommend wrapping in `Variable(-, requires_grad=True)` only in the case that the original variables do not require grad? Or is there a better practice we can follow?

apaszke replied:
A better practice would be to return `self.concentration` and let the caller decide if they want to take grads or not

fritzo replied:
Regardless of whether the user wants to take grads, `.entropy()` and `.kl_divergence()` need to use `grad()` under the hood; that is why we are wrapping this with `requires_grad=True`. That is, even if the user does not want to compute gradients of `Dirichlet(concentration).entropy()`, the Bregman divergence implementation uses `grad()` under the hood and must use a differentiable form of `concentration`, i.e. even if the result of `.entropy()` is not differentiable wrt the user's input. See implementation of `ExponentialFamily.entropy()` for clarification.

apaszke replied:
I understand. I'm just saying that setting `requires_grad=True` should be a responsibility of the caller. If a particular implementation of `entropy()` needs it, then it needs to set the flag.

fritzo replied:
Ok, if I understand correctly then we should simply fail in the following case?
```py
dist = Dirichlet(Variable(torch.ones(2), requires_grad=False))
Dirichlet.entropy(dist)          # Succeeds.
ExponentialFamily.entropy(dist)  # Should this succeed or fail?
```
@vishwakftw I think in this case we can simply add `requires_grad=True` to all the test examples? And maybe add a check in `.entropy()` and `.kl_divergence()`:
```py
nparams = self.natural_params
if not all(p.requires_grad for p in nparams):
    raise NotImplementedError(".entropy() requires all parameters to requires_grad")
```

apaszke replied:
No, I meant this:
```python3
class ExponentialFamily(Distribution):
  def entropy(self):
    natural_params = [Variable(p, requires_grad=True) for p in self.natural_params]
    ...
```

vadimkantorov replied:
If this is frequent (setting `require_grad = True` on existing variables to enable `grad()`, how about adding a convenience method `attach()`? (that was also my suggestion in https://github.com/pytorch/pytorch/issues/1734)

vishwakftw committed:
Major Refactor of the ExponentialFamily system

1. Shift Autograd Semantics to within ExponentialFamily alone, hence making the natural params a Variable is done within ExponentialFamily and is not public.
2. In favour of 1., make log_normalizer a callable property. Calls like self.log_normalizer(*nparams) are valid now.
3. Remove unnecessary imports caused by the removal of autograd semantics from the distributions itself.
4. Modify tests in a minor way
5. Remove ExponentialFamily supports for Binomial and Pareto because both these have 1 out of 2 possible natural parameters (one is fixed)

vishwakftw commented:
After the discussion between @apaszke and @fritzo , I have refactored the system a bit. Please have a look at the commit message for more details.

fritzo commented on torch/distributions/exp_family.py:
```diff
@@ -0,0 +1,80 @@
+        """
+        raise NotImplementedError
+
+    def log_normalizer(self):
```
I think this now needs to be a function
```py
class ExponentialFamily(Distribution):
    def log_normalizer(self, *params):
        raise NotImplementedError

class Exponential(ExponentialFamily):
    def log_normalizer(self, x):
        return -torch.log(-x)
```

vishwakftw replied:
Does the function signature need to be the same? I changed it without changing the signature in `ExponentialFamily` and it still seems to be fine.

fritzo replied:
I think `log_normalizer(self, *natural_params)` is not needed but would clarify how it should be implemented and would serve as documentation for developers.

apaszke replied:
Is it expected that this function has a different signature across distributions? Does the number of arguments match the number of natural params?

fritzo commented on torch/distributions/exponential.py:
```diff
@@ -39,3 +40,12 @@ def log_prob(self, value):
+        V1 = -self.rate.data
+        return (V1, )
+
+    @property
```
A property that returns a lambda is just a function :smile:

vishwakftw replied:
I realized that after I pushed, because this is redundant.

fritzo commented on torch/distributions/exp_family.py:
```diff
@@ -0,0 +1,80 @@
+        raise NotImplementedError
+
+    @property
+    def mean_carrier_measure(self):
```
Is it the case that `mean_carrier_measure` never depends on params? If it never depends on params, then this implementation works fine. However if it ever does depend on params, then you should change this to a function with the same signature as `log_normalizer()`.

vishwakftw replied:
Expected Carrier measure is independent of natural params.

vishwakftw replied:
At least for the distributions we have now, `mean_carrier_measure` is independent of `natural_params`.

fritzo reviewed (commented):
Thanks for refactoring!

vishwakftw committed:
Remove redundancy by eliminating lambda from the log_normalizers

vishwakftw committed:
Modify the entropy for mathematical correctness; this doesn't break any implementation

fritzo commented:
@vishwakftw I kind of like Adam's suggestion of making `._natural_params`, `._log_normalizer()` and `.mean_carrier_measure` private, and keeping only `.entropy()` and `.kl_divergence()` public. What do you think? For the same reason, should we move `.kl_divergence()` from a method to a registered KL implementation?
```py
@register_kl(ExponentialFamily, ExponentialFamily)
def _kl_expfamily_expfamily(p, q):
    if type(p) is not type(q):
        raise NotImplementedError
    ...your implementation here...
```
This would prevent users from accidentally using `ExponentialDistribution.kl_divergence()` outside of testing code; in tests we could directly reference `_kl_expfamily_expfamily()`. WDYT?

vishwakftw committed:
Remove .kl_divergence() from ExponentialFamily and register in kl.py

vishwakftw commented:
I think this is fine w.r.t. KL-Divergences, but what can we do about `.entropy`?

apaszke reviewed (commented):
This looks good. I have three main questions:
1. What's the signature of `log_normalizer` (see comment inline)
2. Are you sure that this API will be stable, or is it an experimental thing mainly for test purposes? If e.g. there exist distributions where the way `mean_carrier_measure` is not implemented this way, we might want to change the signature.
3. Why does `natural_params` return tensors and not variables?

vishwakftw commented:
@apaszke Thanks for the questions:
1. The function signature for `log_normalizer` is `(self, *natural_params)`. Yes, the number of arguments is exactly the number of natural parameters characterizing the distribution. I will modify default implementation in `ExponentialFamily` to match this, as suggested by @fritzo 
2. Theoretically, the mean carrier measure could be a function of natural parameters. But so far, in all the distributions we have seen, this is not the case, and has either been 0 or doesn't have a closed form (which is when we leave it as it is).
3. `natural_params` could return the same type as the actual parameters for the distribution, but this will cause some case-checking and rewrapping in the entropy and KL-divergences, when we have to use the AD system for compute Bregman Divergences.

Let me know your views about this, and clarifications if any. If this is clear, then I will proceed to modify the default function signature. If you have anything you would like me to add in the code, I will make that addition as well. I hope I have addressed your questions well.

fritzo commented on torch/distributions/kl.py:
```diff
@@ -207,6 +208,24 @@ def _kl_exponential_exponential(p, q):
+@register_kl(ExponentialFamily, ExponentialFamily)
+def _kl_expfamily_expfamily(p, q):
+    if not type(p) == type(q):
+        raise ValueError("The cross KL-divergence between different exponential families cannot \
```
I believe this should be a `NotImplementedError`: it is a shortcoming of our implementation rather than a user error due to poorly constructed data.

vishwakftw committed:
Add functions signature for log_normalizer and change exception type in _kl_expfamily_expfamily

apaszke commented:
Thanks @vishwakftw. 
2. But even if that will be the case, it should be still possible to compute the mean carrier measure, because it could just be a property, right?
3. I still think that returning Variables is safer, and we will be able to remove most checking code once Variables and Tensors are merged. You could even just have an assert that checks that args aren't tensors.

vishwakftw committed:
Minor Refactoring:
1. Modify tests to have only Variables and no Tensors
2. Change the return value types for natural_params from Tensor to Variable

vishwakftw commented:
@apaszke Following your suggestions, I have generalized the code to take only `Variable` as input. I have not added any asserts for that however. 

Regarding the mean carrier measure, existing literature so far suggests that it is either 0 or intractable to compute. Maybe @fritzo and @alicanb can weigh in on this.

apaszke commented on torch/distributions/exponential.py:
```diff
@@ -39,3 +40,11 @@ def log_prob(self, value):
+    @lazy_property
+    def natural_params(self):
+        V1 = -self.rate
+        return (V1, )
```
Can you simplify those methods to one liners like `(-self.rate,)`? Also, do we really want them to be lazy properties? They will hold on to their results, consuming more memory than necessary.

apaszke reviewed (commented):
I have some general suggestions for how `natural_params` are implemented (only applies in those extremely simple cases, e.g. `Normal` is fine). I'd still make them regular `@property` in all classes to save memory.

Once this is addressed and conflicts are fixed, it should be good to merge!

fritzo commented:
@vishwakftw I still like Adam's suggestion of making `._natural_params`, `._log_normalizer()` and `._mean_carrier_measure` private, since they are only used to implement generic methods. Mostly I don't want to confuse users of `Normal` with internal-only methods, since that detracts attention from important methods like `.rsample()` and `.entropy()`. Does that sound reasonable?

vishwakftw commented:
@fritzo I think that is a good idea. Especially, because of the fact that natural parameters will be occasionally used, I think it would be better to abstract it away.

We can come with a similar logic as done for `kl_divergence` for these methods as well. Please let me know what you think.

fritzo commented:
@vishwakftw can we just add underscores to the three internal `ExponentialFamily` methods?

vishwakftw committed:
Minor refactor:
1. Make the natural_params, log_normalizer and mean_carrier_measure private
2. Remove lazy_property for memory saving and add returns directly

fritzo reviewed (approved):
Looks great!

vishwakftw committed:
Merge branch 'master' into exp-family

vishwakftw committed:
Fix lint

vishwakftw committed:
Add Poisson distribution in the ExponentialFamily

vishwakftw committed:
Merge branch 'master' into exp-family

apaszke commented on test/test_distributions.py:
```diff
@@ -250,8 +252,8 @@ def is_all_nan(tensor):
-            'low': torch.Tensor([1.0, 1.0]),
-            'high': torch.Tensor([2.0, 3.0]),
+            'low': variable([1.0, 1.0]),
+            'high': variable([2.0, 3.0]),
```
any reason why these two are `variable` instead of `Variable`?

vishwakftw replied:
It is a constructor difference, right? I thought it would be unnecessary to wrap `[1.0, 1.0]` with a `Tensor` and then with `Variable`.

However, to maintain consistency, I am changing this. Thanks for pointing this out.

apaszke commented on test/test_distributions.py:
```diff
@@ -1514,7 +1517,7 @@ def test_entropy_shape(self):
                         Dist.__name__, i + 1, len(params), expected_shape, actual_shape)
                     self.assertEqual(actual_shape, expected_shape, message=message)
-                except NotImplementedError:
+                except (NotImplementedError, ValueError):
```
Why `ValueError`?

vishwakftw replied:
Sorry about that. That was there when the `.entropy` in `ExponentialFamily` returned a `ValueError`. I have removed it now. Thank you.

apaszke commented on torch/distributions/kl.py:
```diff
@@ -210,6 +211,24 @@ def _kl_exponential_exponential(p, q):
+    for pnp, qnp, g in zip(p_nparams, q_nparams, gradients):
+        term = (qnp - pnp) * g
+        for _ in range(len(q.event_shape)):
+            term = term.sum(-1)
```
@fritzo added a helper to compute this sum more efficiently in another PR. I'll merge that one first, and then please rebase and update this code

vishwakftw replied:
Sure. Thank you.

apaszke reviewed (commented):
Blocked on #4937

vishwakftw committed:
Minor changes:
1. Change variable to Variable in EXAMPLES
2. Remove unnecessary ValueError in test_entropy_shape

vishwakftw committed:
Merge branch 'exp-family' of github.com:probtorch/pytorch into exp-family

vishwakftw committed:
Merge branch 'master' into exp-family

vishwakftw committed:
Add _sum_rightmost for KL(ExponentialFamily, ExponentialFamily)

vishwakftw commented:
@apaszke I have modified the kl computation using @fritzo's helper function.

fritzo commented on torch/distributions/kl.py:
```diff
@@ -210,6 +212,22 @@ def _kl_exponential_exponential(p, q):
+    result = q._log_normalizer(*q_nparams) - lg_normal.clone()
+    for pnp, qnp, g in zip(p_nparams, q_nparams, gradients):
+        term = (qnp - pnp) * g
+        result -= _sum_rightmost(term, len(q.event_shape))
```
Nice, thanks for correctly handling the multivariate case.

vishwakftw committed:
Improve raise statement

apaszke commented:
Thanks @vishwakftw!
